gum/arch-mips/gummipsbacktracer.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (C) 2015 Ole André Vadla Ravnås <oleavr@nowsecure.com>
+ *
+ * Licence: wxWindows Library Licence, Version 3.1
+ */
+
+#include "gummipsbacktracer.h"
+
+#include "guminterceptor.h"
+#include "gummemorymap.h"
+
+struct _GumMipsBacktracerPrivate
+{
+  GumMemoryMap * code;
+  GumMemoryMap * writable;
+};
+
+static void gum_mips_backtracer_iface_init (gpointer g_iface,
+    gpointer iface_data);
+static void gum_mips_backtracer_dispose (GObject * object);
+static void gum_mips_backtracer_generate (GumBacktracer * backtracer,
+    const GumCpuContext * cpu_context,
+    GumReturnAddressArray * return_addresses);
+
+G_DEFINE_TYPE_EXTENDED (GumMipsBacktracer,
+                        gum_mips_backtracer,
+                        G_TYPE_OBJECT,
+                        0,
+                        G_IMPLEMENT_INTERFACE (GUM_TYPE_BACKTRACER,
+                                               gum_mips_backtracer_iface_init));
+
+static void
+gum_mips_backtracer_class_init (GumMipsBacktracerClass * klass)
+{
+  GObjectClass * object_class = G_OBJECT_CLASS (klass);
+
+  g_type_class_add_private (klass, sizeof (GumMipsBacktracerPrivate));
+
+  object_class->dispose = gum_mips_backtracer_dispose;
+}
+
+static void
+gum_mips_backtracer_iface_init (gpointer g_iface,
+                                 gpointer iface_data)
+{
+  GumBacktracerIface * iface = (GumBacktracerIface *) g_iface;
+
+  iface->generate = gum_mips_backtracer_generate;
+}
+
+static void
+gum_mips_backtracer_init (GumMipsBacktracer * self)
+{
+  self->priv = G_TYPE_INSTANCE_GET_PRIVATE (self, GUM_TYPE_MIPS_BACKTRACER,
+      GumMipsBacktracerPrivate);
+
+  self->priv->code = gum_memory_map_new (GUM_PAGE_EXECUTE);
+  self->priv->writable = gum_memory_map_new (GUM_PAGE_WRITE);
+}
+
+static void
+gum_mips_backtracer_dispose (GObject * object)
+{
+  GumMipsBacktracer * self = GUM_MIPS_BACKTRACER (object);
+  GumMipsBacktracerPrivate * priv = self->priv;
+
+  if (priv->code != NULL)
+  {
+    g_object_unref (priv->code);
+    priv->code = NULL;
+  }
+
+  if (priv->writable != NULL)
+  {
+    g_object_unref (priv->writable);
+    priv->writable = NULL;
+  }
```

Comment:
For newly written code we should try to use `g_clear_object()` (and `g_clear_pointer()` for non-GObject pointers). E.g.:

```
g_clear_object (&priv->code);

g_clear_object (&priv->writable);
```